File: Difficulty: Easy/Balanced Tree Check/balanced-tree-check.py
class Solution:
    def isBalanced(self, root):
        def check(node):
            if not node:
                return 0  # height of empty tree
            
            left = check(node.left)
            if left == -1:
                return -1
            
            right = check(node.right)
            if right == -1:
                return -1
            
            if abs(left - right) > 1:
                return -1  # unbalanced
            
            return max(left, right) + 1  # height of current node
        
        return check(root) != -1

File: Difficulty: Easy/Balanced Tree Check/balanced-tree-check.cpp
/*
Definition for Node
class Node {
  public:
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = right = nullptr;
    }
};
*/

class Solution {
  public:
  int height(Node* root,bool &valid){
      if(!root) return 0;
      int l=height(root->left,valid);
      int r=height(root->right,valid);
      if(abs(l-r)>1){
          valid=0;
      }
      return 1+max(l,r);
  }
    bool isBalanced(Node* root) {
        // code here
        bool valid=1;
        height(root,valid);
        return valid;
    }
};